backend/src/main/kotlin/cz/loono/backend/api/service/PreventionService.kt

```diff
@@ -36,48 +42,91 @@ class PreventionService(
+        val examsOfType = examinationTypesToRecords[examinationInterval.examinationType]
+        val sortedExamsOfType = examsOfType
+            ?.filter { it ->
+                it.plannedDate != null ||
+                    it.status != ExaminationStatusDto.CONFIRMED ||
+                    it.status != ExaminationStatusDto.CANCELED
+            }
+            ?.sortedBy(ExaminationRecord::plannedDate) ?: listOf(ExaminationRecord())
+
+        val confirmedExamsOfCurrentType = examsOfType?.filter { it.status == ExaminationStatusDto.CONFIRMED }
+        // 1) Filter all the confirmed records
+        // 2) Map all non-nullable lastExamination records
+        // 3) Find the largest or return null if the list is empty
+        val lastConfirmedDate = confirmedExamsOfCurrentType?.mapNotNull(ExaminationRecord::plannedDate)?.maxOrNull()
+        val totalCountOfConfirmedExams = confirmedExamsOfCurrentType?.size ?: 0
+
+        ExaminationPreventionStatusDto(
+            uuid = sortedExamsOfType[0].uuid,
+            examinationType = examinationInterval.examinationType,
+            intervalYears = examinationInterval.intervalYears,
+            plannedDate = sortedExamsOfType[0].plannedDate,
+            firstExam = sortedExamsOfType[0].firstExam,
+            priority = examinationInterval.priority,
+            state = sortedExamsOfType[0].status,
+            count = totalCountOfConfirmedExams,
+            lastConfirmedDate = lastConfirmedDate
+        )
+    }
+
+    private fun prepareSelfExaminationsStatuses(account: Account): List<SelfExaminationPreventionStatusDto> {
+        val result = mutableListOf<SelfExaminationPreventionStatusDto>()
+        val selfExams = selfExaminationRecordRepository.findAllByAccount(account)
+        SelfExaminationTypeDto.values().forEach { type ->
+            val filteredExams = selfExams.filter { exam -> exam.type == type }
+            val suitableSelfExam = validateSexPrerequisities(type, account.userAuxiliary.sex)
+            if (filteredExams.isNotEmpty() && suitableSelfExam) {
+                val plannedExam = filteredExams.filter { exam -> exam.status == SelfExaminationStatusDto.PLANNED }[0]
+                result.add(
+                    SelfExaminationPreventionStatusDto(
+                        lastExamUuid = plannedExam.uuid,
+                        plannedDate = plannedExam.dueDate,
+                        type = type,
+                        history = filteredExams.map { it.status }
+                    )
+                )
+            } else if (suitableSelfExam) {
+                result.add(
+                    SelfExaminationPreventionStatusDto(
+                        type = type,
+                        history = emptyList()
+                    )
+                )
+            }
+        }
+        return result
+    }
+
+    private fun validateSexPrerequisities(type: SelfExaminationTypeDto, sex: String?): Boolean {
+        return when (type) {
```

Comment:
1) Expression here? 
2) Btw, typo `Prerequisites`? 
3) `sex` is nullable, shall we have default branch with `throw` then, when `null` is passed?
private fun validateSexPrerequisities(type: SelfExaminationTypeDto, sex: String?) =
       when (type) {
            SelfExaminationTypeDto.BREAST -> sex == SexDto.FEMALE.name
            SelfExaminationTypeDto.TESTICULAR -> sex == SexDto.MALE.name
        }

Reply:
solved, the account has to be refactored because of all null-checks and issues. discussed during hacketon.